### PR TITLE
feat: composite id backend support

### PR DESCRIPTION
**Pull request by yelhouti**
---

Please make sure the below checklist is followed for Pull Requests.

- [X] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [X] Tests are added where necessary
- [X] The JDL part is updated if necessary
- [X] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [X] Documentation is added/updated where necessary
- [X] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by yelhouti** [comment]
fixes #17919

File: generators/database-changelog-liquibase/templates/src/main/resources/config/liquibase/fake-data/table_entity.csv.ejs
Comment (Tcharl) [inline generators/database-changelog-liquibase/templates/src/main/resources/config/liquibase/fake-data/table_entity.csv.ejs:55]: There are also derived properties for fieldTypeBytes and fieldTypeBlobContentText

File: generators/entity-client/files-angular.cjs
Comment (Tcharl) [inline generators/entity-client/files-angular.cjs:180]: Not quite happy with this one, please remove one or the other parameter

File: generators/entity-server/templates/partials/save_template.ejs
Comment (Tcharl) [inline generators/entity-server/templates/partials/save_template.ejs:46]: Derived property? i.e. primaryKeyComposite

File: generators/entity-client/files-angular.cjs
Comment (yelhouti) [reply generators/entity-client/files-angular.cjs:180]: @Tcharl, thanks a lot for taking the time to review this.
I need to pass the entity for some recursive code. the ...entity, is used everywhere else to be able to access fields, relationships... directly without `entity.X` . I really think this is the best approach, but maybe you have a better suggestion

File: generators/entity-server/templates/partials/save_template.ejs
Comment (Tcharl) [inline generators/entity-server/templates/partials/save_template.ejs:54]: Derived property? i.e. primaryKeyComposite

File: generators/entity-server/templates/partials/update_template.ejs
Comment (Tcharl) [inline generators/entity-server/templates/partials/update_template.ejs:44]: Derived property? i.e. primaryKeyComposite

File: generators/entity-server/templates/partials/update_template.ejs
Comment (Tcharl) [inline generators/entity-server/templates/partials/update_template.ejs:56]: Derived property? i.e. primaryKeyComposite

File: generators/entity-server/templates/src/main/java/package/domain/Entity.java.jhi.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/domain/Entity.java.jhi.ejs:111]: primaryKey.composite

File: generators/entity-server/templates/src/main/java/package/domain/Entity.java.jhi.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/domain/Entity.java.jhi.ejs:196]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/service/mapper/EntityMapper.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/service/mapper/EntityMapper.java.ejs:82]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:27]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:145]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:178]: primaryKey.containsUniqueField?

Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:178]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:162]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:145]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:185]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:194]: prefer use map().orElseGet()

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:253]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:260]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:269]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:322]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:336]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:342]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:346]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:351]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:451]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:462]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:463]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:474]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:482]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:483]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:499]: !primaryKeyTypeString

File: generators/entity-server/templates/src/test/java/package/service/dto/EntityDTOTest.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/test/java/package/service/dto/EntityDTOTest.java.ejs:34]: relationshipTypeOneToOne

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:58]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:62]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:204]: Update fieldsContainsUUID and add derived primaryKeyContainsString (not sure primaryKeyTypeString is useful anymore)

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
Comment (Tcharl) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:737]: primaryKey.containsUniqueField?

**Comment by mraible** [comment]
Conflicts need to be resolved.

**Comment by DanielFran** [comment]
@yelhouti Are you still committed to working on this feature?

**Comment by yelhouti** [comment]
Yes, If anyone is willing to Review and merge.

**Comment by DanielFran** [comment]
@yelhouti you need ist to fix the conflicts

**Comment by yelhouti** [comment]
@DanielFran I did that a few times and as you can imagine, with this big of PR it takes time. As soon as someone volontiers for merging I can do it again. I can also pay one of the maintainers if they want for their time to implement this correctly for the backend. @mshima interested ?

**Comment by yelhouti** [comment]
@DanielFran I fixed the conflicts (if you could please have a look) I still have some small tests to fix ... I m willing to put more time if you are willing to merge

**Comment by yelhouti** [comment]
@mshima thanks a lot man for the work you did on vscode to generate integration tests locally. It was always a PITA to know which test should be run how. Now it all works great and I can fix tests 10 times faster. Maybe this should be documented for others in [CONTRIBUTING.md](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md#running-integration-tests-locally) .

**Comment by yelhouti** [comment]
@mshima @DanielFran finally tests are passing here and in my compositeKey projects. Could anyone please merge this

**Review by mshima — CHANGES_REQUESTED** [review]
Entity suffix is not correct.
I need to do a deeper review.

**Comment by yelhouti** [comment]
@mshima sure, if you can point me to what to fix, I will gladly do

**Comment by mshima** [comment]
At least it looks wrong at generated sample.
IMG_0276

**Comment by yelhouti** [comment]
> At least it looks wrong at generated sample. IMG_0276

When using MapsId, the id of the parent and the child are the same. Put I can change this, if you want, hopefully it won't break the composite with multi level... I must admit I don't remember  why i changed it

**Comment by yelhouti** [comment]
Guess, I need to rebase again ...

**Review by mshima — CHANGES_REQUESTED** [review]
We should avoid complicating templates.

This is a partial review, would be much better if split into multiple PRs.

**Pull request closed without merging**
